### ISSUE TITLE
'kubectl get --all-namespaces pods/xxx' fail

### ISSUE THREAD

**Issue by gouyang** [kubernetes/kubernetes#25403]
`kubectl get --all-namespaces pods` works but `kubectl get --all-namespaces pods/xxx` fail, is it by design or it's a problem?

**Comment by fabioy** [comment]
If the intent was to get pod with name "xxx", then I think the syntax should be:

`kubectl get --all-namespaces pod xxx`

**Comment by janetkuo** [comment]
Neither `kubectl get --all-namespaces pod/xxx` nor `kubectl get --all-namespaces pod xxx` will work. Retrieving pods with specific name from all namespaces isn't supported currently. 

If you wish to retrieve a pod under a namespace, use `kubectl get pod/xxx --namespace=xxx`

**Comment by nhlfr** [comment]
I'm going to work on this issue and bring the support for retrieving pods by name from all namespaces - just FYI.

**Comment by nhlfr** [comment]
From what I see, it should be more an API change than only kubectl. Why? Because IMO the issue comes from the inconsistency between "get list" and "get the single element" actions in relation to namespace.

List operations:

* */api/v1/pods* - it gets the pods from the **all** namespaces - it's what `kubectl get --all-namespaces pods` calls
* */api/v1/namespaces/default/pods*  - it gets the pods from the **default** namespace - it's what `kubectl get pods` calls, because kubectl is using the default namespace explicitly in this case
* */api/v1/namespaces/\<namespace\>/pods* - that's what `kubectl get --namespace=<namespace> pods` is doing

Single element operations:

* */api/v1/pods/<name>* - doesn't exist and has to be implemented - at least I see implementing it as a solution of this bug
* */api/v1/namespaces/default/pods/\<name\>* - that's what `kubectl get pod <name>` is doing
* */api/v1/namespaces/\<namespace\>/pods/\<name\>* - that's what `kubectl get --namespace=<namespace> get pod <name>` is doing

tl;dr the solution is to support */api/v1/\<resource\>/\<name\>* as something which gets the resource by name from all namespaces

**Comment by kargakis** [comment]
@kubernetes/kubectl is this something we want to fix? I don't see an issue here.

**Comment by deads2k** [comment]
> tl;dr the solution is to support /api/v1/<resource>/<name> as something which gets the resource by name from all namespaces

I think that names are naturally namespace scoped, so specifying a name to retrieve across namespaces would result in weird API uses where doing a get by name results in a list.

In addition, I don't think that I want the API server to change to make directly spanning list requests.  Its not impossible to do if you're willing to hold the entire dataset in memory and re-index the entire thing, but the data isn't stored in a way that makes that change efficient.  In addition, namespaces are natural ACL boundaries (see both the abac and rbac authorizers).   This would be an attempt to span that gap and wouldn't play nicely with authorization evaluation (basically everyone would be denied).

**Comment by smarterclayton** [comment]
Not supporting this was by design - as david mentioned, users who want to fetch from multiple namespaces need to do so manually, which could be extremely expensive.  Part of the reason for that (beyond the reasons david mentioned) is that namespaces are *shard boundaries* and we reserve the right in the future to make the "all-namespaces" call have to fetch data from multiple underlying shards.  As a consequence of that, we deliberately want to limit the the scope of cross namespace API calls we expose to preserve the ability to shard at some point in the future.

**Comment by nhlfr** [comment]
So, are we going to close this issue?

**Comment by dims** [comment]
@janetkuo @nhlfr +1 to close this issue

**Comment by AdoHe** [comment]
@dims @pwittrock feel free to close this.
